File: ts/indicators/john-ehlers/hilbert-transformer/cycle-estimator.ts
/** Describes a common Hilbert transformer cycle estimator functionality. */
export interface HilbertTransformerCycleEstimator {
  /** The underlying linear-Weighted Moving Average (WMA) smoothing length. */
  readonly smoothingLength: number; // integer

  /** The current WMA-smoothed value used by underlying Hilbert transformer.
   * 
   * The linear-Weighted Moving Average has a window size of __smoothingLength__.
   */
  readonly smoothed: number;

  /** The current de-trended value. */
  readonly detrended: number;

  /** The current Quadrature component value. */
  readonly quadrature: number;

  /** The current InPhase component value. */
  readonly inPhase: number;

  /** The current period value. */
  readonly period: number;

  /** The current count value. */
  readonly count: number; // integer

  /** Indicates whether an estimator is primed. */
  readonly primed: boolean;

  /** The minimal cycle period supported by this Hilbert transformer. */
  readonly minPeriod: number; // integer

  /** The maximual cycle period supported by this Hilbert transformer. */
  readonly maxPeriod: number; // integer

  /** The value of α (0 < α ≤ 1) used in EMA to smooth the in-phase and quadrature components. */
  readonly alphaEmaQuadratureInPhase: number;

  /** The value of α (0 < α ≤ 1) used in EMA to smooth the instantaneous period. */
  readonly alphaEmaPeriod: number;

  /** The number of updates before the estimator is primed (MaxPeriod * 2 = 100). */
  readonly warmUpPeriod: number; // integer

  /** Updates the estimator given the next sample value. */
  update(sample: number): void;
}

File: go/indicators/johnehlers/hilberttransformer/cycleestimator.go
package hilberttransformer

// CycleEstimator describes a common Hilbert transformer cycle estimator functionality.
type CycleEstimator interface { //nolint:interfacebloat
	// SmoothingLength returns the underlying WMA smoothing length in samples.
	SmoothingLength() int

	// Smoothed returns the current WMA-smoothed value used by underlying Hilbert transformer.
	//
	// The linear-Weighted Moving Average has a window size of SmoothingLength.
	Smoothed() float64

	// Detrended returns the current detrended value.
	Detrended() float64

	// Quadrature returns the current Quadrature component value.
	Quadrature() float64

	// InPhase returns the current InPhase component value.
	InPhase() float64

	// PeriodValue returns the current period value.
	Period() float64

	// Count returns the current count value.
	Count() int

	// Primed indicates whether an instance is primed.
	Primed() bool

	// MinPeriod returns the minimal cycle period.
	MinPeriod() int

	// MaxPeriod returns the maximual cycle period.
	MaxPeriod() int

	// AlphaEmaQuadratureInPhase returns the value of α (0 < α ≤ 1)
	// used in EMA to smooth the in-phase and quadrature components.
	AlphaEmaQuadratureInPhase() float64

	// AlphaEmaPeriod returns the value of α (0 < α ≤ 1)
	// used in EMA to smooth the instantaneous period.
	AlphaEmaPeriod() float64

	// WarmUpPeriod returns the number of updates before the estimator
	// is primed (MaxPeriod * 2 = 100).
	WarmUpPeriod() int

	// Update updates the estimator given the next sample value.
	Update(sample float64)
}
